kernel/src/table_changes/mod.rs

```diff
@@ -0,0 +1,196 @@
+///
+/// [`CommitInfo`]: crate::actions::CommitInfo
+#[derive(Debug)]
+pub struct TableChanges {
+    pub log_segment: LogSegment,
+    table_root: Url,
+    end_snapshot: Snapshot,
+    start_version: Version,
+    schema: Schema,
+}
+
+impl TableChanges {
+    /// Creates a new [`TableChanges`] instance for the given version range. This function checks
+    /// these two properties:
+    /// - The change data feed table feature must be enabled in both the start or end versions.
+    /// - The schema at the start and end versions are the same.
+    ///
+    /// Note that this does not check that change data feed is enabled for every commit in the
+    /// range. It also does not check that the schema remains the same for the entire range.
+    ///
+    /// # Parameters
+    /// - `table_root`: url pointing at the table root (where `_delta_log` folder is located)
+    /// - `engine`: Implementation of [`Engine`] apis.
+    /// - `start_version`: The start version of the change data feed
+    /// - `end_version`: The end version of the change data feed. If this is none, this defaults to
+    ///   the newest table version.
+    pub fn try_new(
+        table_root: Url,
+        engine: &dyn Engine,
+        start_version: Version,
+        end_version: Option<Version>,
+    ) -> DeltaResult<Self> {
+        let start_snapshot =
+            Snapshot::try_new(table_root.as_url().clone(), engine, Some(start_version))?;
+        let end_snapshot = Snapshot::try_new(table_root.as_url().clone(), engine, end_version)?;
+
+        // Verify CDF is enabled at the beginning and end of the interval
+        let is_cdf_enabled = |snapshot: &Snapshot| {
+            static ENABLE_CDF_FLAG: &str = "delta.enableChangeDataFeed";
+            snapshot
+                .metadata()
+                .configuration
+                .get(ENABLE_CDF_FLAG)
+                .is_some_and(|val| val == "true")
+        };
+        if !is_cdf_enabled(&start_snapshot) {
+            return Err(Error::table_changes_disabled(start_version));
+        } else if !is_cdf_enabled(&end_snapshot) {
+            return Err(Error::table_changes_disabled(end_snapshot.version()));
+        }
+        if start_snapshot.schema() != end_snapshot.schema() {
+            return Err(Error::generic(format!(
+                "Failed to build TableChanges: Start and end version schemas are different. Found start version schema {:?} and end version schema {:?}", start_snapshot.schema(), end_snapshot.schema(),
+            )));
+        }
+
+        let log_root = table_root.join("_delta_log/")?;
+        let log_segment = LogSegment::for_table_changes(
+            engine.get_file_system_client().as_ref(),
+            log_root,
+            start_version,
+            end_version,
+        )?;
+
+        let schema = StructType::new(
+            end_snapshot
+                .schema()
+                .fields()
+                .cloned()
+                .chain(CDF_FIELDS.clone()),
```

Comment:
The CDF fields are generated columns right? (not read directly from parquet)?

Reply:
Yes, correct. I was planning on making them a special case only in CDF code.  If you feel that we can legitimately treat these as generated column, we could add a new column type `ColumnType::GeneratedColumn`.